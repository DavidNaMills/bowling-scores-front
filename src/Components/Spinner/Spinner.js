import React from "react";
import { css } from "@emotion/core";
import { PropagateLoader } from "react-spinners";

const centerStyle = {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    height: '20vh',
}

// Can be a string as well. Need to ensure each key-value pair ends with ;
const override = css`
  display: block;
  margin: 0 auto;
`;

const defColor = '36, 69, 255'

const Spinner = ({ size = 20, css = override, color = defColor, center=false }) => {
    return (
        <div style={center ? centerStyle : null}>

            <PropagateLoader
                css={css}
                size={size}
                color={`rgb(${color})`}
                loading={true}
            />
        </div>
    );
}

export default Spinner;