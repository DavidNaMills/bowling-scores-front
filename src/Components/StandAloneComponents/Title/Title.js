import React from 'react';
import classes from './Title.module.scss';

const Title = ({ label = '', ttlType = 'sub' }) => {
    let item;

    switch (ttlType.toLowerCase()) {
        case 'main':
            item = <h1 className={classes.title__main}>{label}</h1>
            break;
        case 'section':
            item = <h2 className={classes.title__section}>{label}</h2>
            break;
        case 'sub':
            item = <h3 className={classes.title__sub}>{label}</h3>
            break;
        default:
            item = <h3 className={classes.title__sub}>{label}</h3>
            break;
    }

    return (
        <div className={classes.title}>
            {item}
        </div>
    )
}

export default Title;