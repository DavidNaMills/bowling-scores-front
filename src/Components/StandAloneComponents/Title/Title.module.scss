@import '../../../styles/colors.scss';
@import '../../../styles/fonts.scss';

.title{
    // -webkit-box-shadow: $title_shadow_full;
    // -moz-box-shadow: $title_shadow_full;
    // box-shadow: $title_shadow_full;
    padding: 2px;
    text-align: center;
    font-family: $header;

    &__main{
        font-size: $header_size_main;
        color: $title_main_color;
    }
    
    &__section{
        font-size: $header_size_section;
        color: $title_rest_color;
    }
    &__sub{
        font-size: $header_size_sub;
        color: $title_rest_color;
    }
}