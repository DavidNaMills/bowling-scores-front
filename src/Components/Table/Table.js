import React from 'react';
import classes from './Table.module.scss';

const Table = ({ caption = '', tableClass = classes.table, data, showRowNum = false, selectRow = () => { }, ...rest }) => (
    <div>
        {caption && <p className={classes.caption}>{caption}</p>}
        <table className={tableClass}>
            <thead>
                {showRowNum && <th></th>}
                {data.headers.map((x, i) =>
                    <th key={i}>
                        {x}
                    </th>
                    , 0)}
            </thead>
            <tbody>
                {
                    data.rows.map((x, i) =>
                        <tr
                            key={i}
                            className={x.style && typeof (x.style.trow) === 'string' ? x.style.trow : ''}
                            style={x.style && typeof (x.style.trow) === 'object' ? x.style.trow : {}}
                            onClick={() => selectRow(x.id ? x.id : i + 1)}
                        >
                            {showRowNum && <td key={`${i}0`}>{i + 1}</td>}
                            {
                                x.values.map((y, o) =>
                                    <td key={`${i}-${o}`}>
                                        {y}
                                    </td>
                                    , 1)
                            }
                        </tr>
                    )
                }
            </tbody>
        </table>
    </div>
)

export default Table;

// const tHeaders = ['name', 'Average', 'Pinfall', 'Rank'];
// const tData = [
//     {
//         id: '',
//         values: ['david', 135, 2445, '+ 2'],
//         style: { tr: classes.testRowClass }  //optional
//     },
//     {
//         id: '5vd5f3d5cvx3cv5xz53xcz',
//         values: ['alan', 189, 2222, '-2'],
//         style: { trow: { color: 'red' } }  //optional
//     }
// ];