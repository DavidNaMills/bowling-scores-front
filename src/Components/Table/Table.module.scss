.table{
    width: 100%;
    border: solid black 1px;
    border-collapse: collapse;
    color: black;
    

    thead {
        
        background: rgb(255,255,255);
        background: -moz-linear-gradient(180deg, rgba(255,255,255,1) 37%, rgba(201,201,201,1) 100%);
        background: -webkit-linear-gradient(180deg, rgba(255,255,255,1) 37%, rgba(201,201,201,1) 100%);
        background: linear-gradient(180deg, rgba(255,255,255,1) 37%, rgba(201,201,201,1) 100%);
        filter: progid:DXImageTransform.Microsoft.gradient(startColorstr="#ffffff",endColorstr="#c9c9c9",GradientType=1);
    }

    th{
        font-size: 16px;
        border: solid rgb(111, 111, 111) 1px;
        color: black;
        padding: 5px;
    }
    
    td{
        font-size: 14px;
        border: solid rgb(111, 111, 111) 1px;
        padding: 5px;
        text-align: center;
    }  
}

.caption{
    font-size: 14px;
    padding:0;
    margin-top:15px;
    margin-bottom:5px;
}