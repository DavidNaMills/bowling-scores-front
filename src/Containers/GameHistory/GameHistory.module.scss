@import '../../styles/shared/spacing.module.scss';

@mixin baseShadow() {
    margin-bottom: 70px;
    -webkit-box-shadow: 5px 2px 15px 5px rgba(36,112,107,1);
    -moz-box-shadow: 5px 2px 15px 5px rgba(36,112,107,1);
    box-shadow: 5px 2px 15px 5px rgba(36,112,107,1);
    padding: 5px;
}

@mixin base {
    width: 100%;
    display: flex;
}

.gameHistory{
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    flex-wrap: wrap;

    &__game{
        &_desktop{
            @include baseShadow();
            width: 450px;
        }
    
        &_mobile{
            @include baseShadow();
            width: 100%;
        }
    }

    &__spinner{
        display: flex;
        margin: auto;
        padding-top: 50px;
        padding-bottom: 50px;
    }

}