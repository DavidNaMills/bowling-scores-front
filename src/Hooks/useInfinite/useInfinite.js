import React, { useEffect, useState, useRef, useCallback, useMemo } from 'react';
import classes from './useInfinite.module.scss';

import Spinner from '../../Components/Spinner/Spinner';


const useInfinite = (axios, url, locale) => {

    const [items, setItems] = useState([]);
    const [skip, setSkip] = useState(0);
    const [val, setVal] = useState('');
    const [link, setLink] = useState('?search=');
    const [infiniteLoading, setProgress] = useState(false);
    const totalRef = useRef(0);

    useEffect(() => {
        fetchItems();
    }, [link])


    const refresh = () => {
        const t = items.length;
        setProgress(true);
        axios({
            method: 'GET',
            url: `${url}/0/${t}${link}`
        })
            .then(res => {
                setProgress(false);
                setItems(res.data);
                setSkip(+res.data.length);
            })
            .catch(err => {
                setProgress(false);
                // console.log(err);
            })
    }

    const reset = useCallback(() => {
        setItems([]);
        setSkip(0);
    }, []);


    const searchAll = (e) => {
        const t = e.target.value;
        setVal(t);
        setItems([]);
        setSkip(0);
        let delayTimer;
        clearTimeout(delayTimer);
        delayTimer = setTimeout(function () {
            setLink(`/?search=${t}`);
        }, 1000);
    }


    const fetchItems = (limit = 5) => {
        setProgress(true);
        axios({
            method: 'GET',
            url: `${url}/${skip}/${limit}${link}`
        })
            .then(res => {
                setProgress(false);
                if (res.data.totalCount) {
                    totalRef.current = res.data.totalCount;
                }
                setItems(prev => [...prev, ...res.data.items]);
                setSkip(prev => prev + +res.data.items.length);
                // setProgress(false);
            })
            .catch(err => {
                setProgress(false);
                // console.log(err);
            })
    }


    /**
     * JSX functions
     */

    const loadMoreBar = () => (
        items.length == +totalRef.current
            ? <p className={[classes.Infinite__more, classes.Infinite__more_finished].join(' , ')} >No more games to fetch</p>
            : <p className={[classes.Infinite__more, classes.Infinite__more_toFetch].join(' , ')} onClick={() => fetchItems()}>
                {infiniteLoading
                    ? <span className={classes.Infinite__spinner}><Spinner size={16}/></span>
                    : `load more games?`
                }
            </p>
    )

    const infiniteSearchBar = () => (
        <div className={classes.Infinite}>
            <label>search</label>
            <input value={val} onChange={e => searchAll(e)} className={classes.Infinite__input} />
        </div>
    )

    const searchResultsBar = () => {
        const msg = items.length == +totalRef.current
            ? <p className={classes.Infinite__results}>{`Displaying all games (${totalRef.current})`}</p>
            : <p className={classes.Infinite__results}>{`Displaying ${items.length} ${totalRef.current && `of ${isNaN(totalRef.current) ? 0 : totalRef.current}`} games`}</p>
        return msg
    }



    return {
        items,      //search results

        searchAll,      // search function
        fetchItems,     // standalone fetch items
        reset,          // reset all
        refresh,        // refreshes the current list

        infiniteLoading,    //for spinner

        loadMoreBar,        //jsx to display load more bar
        infiniteSearchBar,  //jsx search bar
        searchResultsBar    //jsx results count bar
    }
};

export default useInfinite;