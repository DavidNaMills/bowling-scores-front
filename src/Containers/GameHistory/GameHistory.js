import React, { useMemo } from 'react';
import { NavLink } from 'react-router-dom';
import { useSelector } from 'react-redux';

import classes from './GameHistory.module.scss';
// import container from '../../styles/shared/container.module.scss';

import GameChart from '../../Components/GameChart/GameChart';
import Table from '../../Components/Table/Table';
import Title from '../../Components/StandAloneComponents/Title/Title';
import Spinner from '../../Components/Spinner/Spinner';
import { isMobileOnly, withOrientationChange } from 'react-device-detect';

import chartParser from '../../helpers/chartDataModifier/chartDataModifier';
import pinfallSort from '../../helpers/pinfallSort/pinfallSort';
import tableParser from '../../helpers/tableGamesDataModifier/tableGamesDataModifier';
import generateDate from '../../helpers/generateDate/generateDate';

import axios from '../../Axios/axiosConfig';
import { PERSONAL_GAMES } from '../../Axios/URLS';
import useInfinite from '../../Hooks/useInfinite/useInfinite';

import Anime, { anime } from 'react-anime';

const tHeaders = ['Player', 'Average', 'Pinfall'];

const toObj = (item) => {
    const t = {};
    for (let i in item) {
        t[i] = {
            id: i,
            ...item[i]
        }
    }
    return t;
}


const GameHistory = (props) => {
    const { items, loadMoreBar, searchResultsBar, infiniteLoading } = useInfinite(axios, PERSONAL_GAMES, {});
    const user = useSelector(state => state.user);
    const placeHolder = () => { };

    const all = useMemo(() =>
        items.map((x, y) =>
            <Anime key={y} delay={anime.stagger(150)}
                scale={[.1, .9]}>
                <div className={isMobileOnly ? classes.gameHistory__game_mobile : classes.gameHistory__game_desktop} key={x._id}>
                    <Title label={generateDate(x.createdAt)} />

                    <GameChart
                        players={toObj(x.players)}
                        data={chartParser(x)}
                        width={isMobileOnly ?
                            props.isPortrait
                                ? (window.innerWidth - 10)
                                : (window.innerWidth - 100)
                            : 440
                        }
                    />

                    <Table
                        data={{
                            headers: tHeaders,
                            rows: pinfallSort(tableParser(x))
                        }}
                        showRowNum
                        selectRow={placeHolder}
                        caption=''
                    />
                </div>
            </Anime>
        ), [items]);


    return (
        <div>
            <Title label='History' ttlType='section' />
            {searchResultsBar()}
            <div className={classes.gameHistory}>
                {
                    items.length > 0
                        ? all
                        : infiniteLoading
                            ? <div className={classes.gameHistory__spinner}><Spinner color={user.user.color} /></div>
                            : <NavLink to='/game'>No games to display. Play one?</NavLink>
                }
            </div>
            {loadMoreBar()}
        </div>
    )
}

export default withOrientationChange(GameHistory);