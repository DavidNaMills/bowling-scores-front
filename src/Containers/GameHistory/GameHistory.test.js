import React from 'react';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';

import { BrowserRouter, NavLink } from 'react-router-dom';
import GameHistory from './GameHistory';
import Title from '../../Components/StandAloneComponents/Title/Title';
import Spinner from '../../Components/Spinner/Spinner';
import Table from '../../Components/Table/Table';


jest.mock('react-redux');
import { useSelector } from 'react-redux';
const mockToken = '1v2d3s1v2s';

// jest.mock('react-device-detect');
// import { isMobileOnly, withOrientationChange } from 'react-device-detect';

jest.mock('../../Hooks/useInfinite/useInfinite');
import useInfinite from '../../Hooks/useInfinite/useInfinite';

import Anime from 'react-anime';


const mockSearchResultsBar = jest.fn(() => <div></div>);
const mockInfiniteSearchBar = jest.fn(() => <div></div>);
const mockLoadMoreBar = jest.fn(() => <div></div>);

const baseFns = {
    fetchItems: jest.fn(() => []),
    searchResultsBar: mockSearchResultsBar,
    infiniteSearchBar: mockInfiniteSearchBar,
    loadMoreBar: mockLoadMoreBar
}

const mockItems = [
    {
        _id: 'fdafdafdas',
        createdAt: '2020-01-21T07:45:51.862+00:00',
        players: {
            '5e26aacc7483c043600de418': {
                name: "super balls",
                color: "3, 169, 244"
            }
        },
        games: {
            "1": {
                '5e26aacc7483c043600de418': {
                    name: "super balls",
                    score: 123
                }
            },
            "2": {
                '5e26aacc7483c043600de418': {
                    name: "super balls",
                    score: 159
                }
            },
        }
    },
    {
        _id: 'fdafdafdas',
        createdAt: '2020-01-21T07:45:51.862+00:00',
        players: {
            '5e26aacc7483c043600de418': {
                name: "super balls",
                color: "3, 169, 244"
            }
        },
        games: {
            "1": {
                '5e26aacc7483c043600de418': {
                    name: "super balls",
                    score: 111
                }
            },
            "2": {
                '5e26aacc7483c043600de418': {
                    name: "super balls",
                    score: 222
                }
            },
        }
    },
]

afterEach(() => {
    jest.clearAllMocks();
});

describe('<GameHistory /> component test suite', () => {
    describe('basic tests', () => {
        it('snapshot tests', () => {
            useSelector.mockImplementation(() => ({
                token: mockToken,
                user: { color: '123, 123, 123' }
            }));
            useInfinite.mockImplementation(() => ({
                ...baseFns,
                items: [],
                infiniteLoading: false,
            }));

            const comp = renderer.create(<BrowserRouter><GameHistory /></BrowserRouter>);
            const app = comp.toJSON();
            expect(app).toMatchSnapshot();
        });

        it('renders a <Title/> component, searchResultsBar and <div> with className="gameHistory"', () => {
            useSelector.mockImplementation(() => ({
                token: mockToken,
                user: { color: '123, 123, 123' }
            }));
            useInfinite.mockImplementation(() => ({
                ...baseFns,
                items: [],
                infiniteLoading: false,
            }));

            const wrapper = mount(<BrowserRouter><GameHistory /></BrowserRouter>);
            expect(wrapper.find(Title).length).toBe(1);
            expect(wrapper.find('.gameHistory').length).toBe(1);
            expect(mockSearchResultsBar).toHaveBeenCalledTimes(1);
            expect(mockLoadMoreBar).toHaveBeenCalledTimes(1);
            useSelector.mockClear();
        });

        it('renders a NavLink if no games and infiniteLoading = false', () => {
            useSelector.mockImplementation(() => ({
                token: mockToken,
                user: { color: '123, 123, 123' }
            }));
            useInfinite.mockImplementation(() => ({
                ...baseFns,
                items: [],
                infiniteLoading: false,
            }));
            const wrapper = mount(<BrowserRouter><GameHistory /></BrowserRouter>);

            expect(wrapper.find(NavLink).length).toBe(1);
            expect(wrapper.find(Spinner).length).toBe(0);
            expect(mockSearchResultsBar).toHaveBeenCalledTimes(1);
            expect(mockLoadMoreBar).toHaveBeenCalledTimes(1);
            useSelector.mockClear();
        });

        it('renders the <Spinner/> component if no games and infiniteLoading = false', () => {
            useSelector.mockImplementation(() => ({
                token: mockToken,
                user: { color: '123, 123, 123' }
            }));
            useInfinite.mockImplementation(() => ({
                ...baseFns,
                items: [],
                infiniteLoading: true,
            }));
            const wrapper = mount(<BrowserRouter><GameHistory /></BrowserRouter>);

            expect(wrapper.find(NavLink).length).toBe(0);
            expect(wrapper.find(Spinner).length).toBe(1);
            expect(mockSearchResultsBar).toHaveBeenCalledTimes(1);
            expect(mockLoadMoreBar).toHaveBeenCalledTimes(1);
            useSelector.mockClear();
        });
    });

    describe('content render tests', () => {
        let conWrap;

        beforeEach(() => {
            useSelector.mockImplementation(() => ({
                token: mockToken,
                user: { color: '123, 123, 123' }
            }));
            useInfinite.mockImplementation(() => ({
                ...baseFns,
                items: mockItems,
                infiniteLoading: false,
            }));
            conWrap = mount(<BrowserRouter><GameHistory /></BrowserRouter>);
        });

        it('renders 2 <Anime/> components', () => {
            expect(conWrap.find(Anime).length).toBe(2);
        });

        it('renders 2 <Table/> components', () => {
            expect(conWrap.find(Table).length).toBe(2);
        });
        
        it('renders 2 divs with class gameHistory__game_desktop if desktop', () => {
            expect(conWrap.find('.gameHistory__game_desktop').length).toBe(2);
        });
    });

    // TODO: test with isMobile
});