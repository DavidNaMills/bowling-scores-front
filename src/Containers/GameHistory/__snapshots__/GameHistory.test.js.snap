// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameHistory /> component test suite basic tests snapshot tests 1`] = `
<div>
  <div
    className="title"
  >
    <h2
      className="title__section"
    >
      History
    </h2>
  </div>
  <div />
  <div
    className="gameHistory"
  >
    <a
      aria-current={null}
      href="/game"
      onClick={[Function]}
    >
      No games to display. Play one?
    </a>
  </div>
  <div />
</div>
`;
